app/grpc/gasestimation/gas_estimator.go

```diff
@@ -105,144 +88,120 @@ func (s *gasEstimatorServer) EstimateGasPriceAndUsage(ctx context.Context, reque
-		ctx,
-		lastFiveBlocksTransactionsQuery(latestHeight),
-		false,
-		&page,
-		&numberOfTransactionsPerPage,
-		"asc",
-	)
+	if txsResp.TotalBytes < int64(appconsts.DefaultMaxBytes*gasPriceEstimationThreshold) {
+		return appconsts.DefaultMinGasPrice, nil
+	}
+	gasPrices, err := SortAndExtractGasPrices(s.txDecoder, txsResp.Txs, int64(appconsts.DefaultUpperBoundMaxBytes))
 	if err != nil {
 		return 0, err
 	}
-
-	totalNumberOfTransactions := txSearchResult.TotalCount
-	if totalNumberOfTransactions == 0 {
-		// return the min gas price if no transaction found in the last 5 blocks
-		return minfee.DefaultNetworkMinGasPrice.MustFloat64(), nil
-	}
-
-	gasPrices := make([]float64, 0)
-	for {
-		currentPageGasPrices, err := extractGasPriceFromTransactions(txSearchResult.Txs)
-		if err != nil {
-			return 0, err
-		}
-		gasPrices = append(gasPrices, currentPageGasPrices...)
-		if len(gasPrices) >= totalNumberOfTransactions {
-			break
-		}
-		page++
-		txSearchResult, err = s.clientCtx.Client.TxSearch(
-			ctx,
-			lastFiveBlocksTransactionsQuery(latestHeight),
-			false,
-			&page,
-			&numberOfTransactionsPerPage,
-			"asc",
-		)
-		if err != nil {
-			return 0, err
-		}
-	}
 	return estimateGasPriceForTransactions(gasPrices, priority)
 }
 
 // estimateGasPriceForTransactions takes a list of transactions and priority
 // and returns a gas price estimation.
 // The priority sets the estimation as follows:
-// - High Priority: The gas price is the price at the start of the top 10% of transactions’ gas prices from the last five blocks.
-// - Medium Priority: The gas price is the mean of all gas prices from the last five blocks.
-// - Low Priority: The gas price is the value at the end of the lowest 10% of gas prices from the last five blocks.
-// - Unspecified Priority (default): This is equivalent to the Medium priority, using the mean of all gas prices from the last five blocks.
+// - High Priority: The gas price is the median price of the top 10% of transactions’ gas prices in the mempool.
+// - Medium Priority: The gas price is the median price of the all gas prices in the mempool.
+// - Low Priority: The gas price is the median price of the bottom 10% of gas prices in the mempool.
+// - Unspecified Priority (default): This is equivalent to the Medium priority, using is the median price of all gas prices in the mempool.
 // More information can be found in ADR-023.
 func estimateGasPriceForTransactions(gasPrices []float64, priority TxPriority) (float64, error) {
-	meanGasPrice := Mean(gasPrices)
 	switch priority {
 	case TxPriority_TX_PRIORITY_UNSPECIFIED:
-		return meanGasPrice, nil
+		return Median(gasPrices)
 	case TxPriority_TX_PRIORITY_LOW:
-		stDev := StandardDeviation(meanGasPrice, gasPrices)
-		return meanGasPrice - EstimationZScore*stDev, nil
+		bottom10PercentIndex := len(gasPrices) * 10 / 100
+		if bottom10PercentIndex == 0 {
+			// the case of a slice containing less than 10 elements
+			bottom10PercentIndex = 1
```

Comment:
[question] what if there are 0 elements then an index of 1 will be out of bounds

Reply:
this expects 70% of the block to be filled, which means the list of gas prices can't be empty. But I guess it makes more sense to add an explicit check:

c9582bce2b719989fa475406ba3ff822f65e4e07